.editorconfig

```diff
@@ -156,6 +156,9 @@ csharp_space_between_square_brackets = false
 # Analyzers
 dotnet_code_quality.ca1802.api_surface = private, internal
 
+# License header
+file_header_template = Licensed to the .NET Foundation under one or more agreements.\nThe .NET Foundation licenses this file to you under the MIT license.\nSee the LICENSE file in the project root for more information.
```

Comment:
How does this behave when we explicitly diverge from this header, e.g. when bringing over ported code with its own license we need to keep and augment?

Reply:
@sharwell based on https://github.com/dotnet/roslyn/issues/33012 I guess are the right person to answer this?

Reply:
The file will need to have this header first, followed by a blank line, and finally any other text can be included. Here's an example of a compatible header:

https://github.com/dotnet/roslyn/blob/d17b2f451a9b603a9a23f541dff4a1c6c25c107f/src/Compilers/Core/Portable/InternalUtilities/NullableAttributes.cs#L1-L6

The second paragraph could have been any notice from the original content.

Reply:
> The file will need to have this header first, followed by a blank line, and finally any other text can be included.

This is going to be problematic, then, as in cases where we deal with code from places like mono that were covered by a different license, we omit the second line of the added header, e.g.
https://github.com/dotnet/runtime/blob/629dba55058b58a126d08fc37db30b0cfa844dde/src/libraries/System.Text.RegularExpressions/tests/MonoRegexTests.cs#L1-L9

Reply:
You can use more specific section(s) in the .editorconfig file to apply a different header to these files.

Reply:
You can also suppress the diagnostic in these files by placing a `#pragma warning disable` on the first line.

Reply:
Can you elaborate?  If there's a way to say "the following headers are all fine, but if none of these are there, use xyz as the header", that'd be fine.

Reply:
A section like this would change the header for the file above:

```
[runtime/src/libraries/System.Text.RegularExpressions/tests/MonoRegexTests.cs]
file_header_template = Licensed to the .NET Foundation under one or more agreements.\nSee the LICENSE file in the project root for more information.
```

With the files spread out, it may be more efficient to just have this at the top of impacted files (with the added advantage that the discrepancy is clearly intentional):

```
#pragma warning disable IDE0073 // File header mismatch (intentional)
```

Reply:
Can the editorconfig support be updated to split what headers are recognized vs what header should be added if none of them are there? e.g.
```C#
file_header_valid.1 = Licensed to the .NET Foundation under one or more agreements.\nThe .NET Foundation licenses this file to you under the MIT license.\nSee the LICENSE file in the project root for more information.
file_header_valid.2 = Licensed to the .NET Foundation under one or more agreements.\nSee the LICENSE file in the project root for more information.
file_header_template = file_header_valid.1
```
or something along those lines?

Reply:
ping @sharwell.

@stephentoub What's your opinion on the feature? Do we want to leverage it as-is or wait for it becoming "more mature"?

Reply:
> Can the editorconfig support be updated to split what headers are recognized vs what header should be added if none of them are there? e.g.

It could in theory. My concern is it further confuses an area which is already difficult for end users to understand. It seems like a project would benefit from one of the following instead:

1. Resolve the discrepancy, such that the source files are consistently licensed throughout.
2. Reorganize the repository, such that source files with the same license appear in the same folders. If a group of files needs to use a different header, those files can be placed into a shared project in a new folder, and the current file header feature used to assign the correct header for that subfolder.

Reply:
Neither of those is a viable option for us.

Reply:
> What's your opinion on the feature? Do we want to leverage it as-is or wait for it becoming "more mature"?

Let's just close this for now.  Thanks.

Reply:
Thanks for the help Sam